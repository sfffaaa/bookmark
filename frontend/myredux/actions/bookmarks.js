import {
    CREATE_BOOKMARK_ACTION_TYPE,
    LIST_BOOKMARK_ACTION_TYPE,
    DELETE_BOOKMARK_ACTION_TYPE,
    UPGRADE_BOOKMARK_ACTION_TYPE,
} from '../constants/action-types';

const BASE_URL = 'http://localhost:3000';

export const createBookmark = url => ({
    type: 'API',
    payload: {
        bookURL: url,
        url: `${BASE_URL}/api/create`,
    },
    actionType: CREATE_BOOKMARK_ACTION_TYPE,
});

export const listBookmark = () => ({
    type: 'API',
    payload: {
        url: `${BASE_URL}/api/list`,
    },
    actionType: LIST_BOOKMARK_ACTION_TYPE,
});

export const deleteBookmark = bookmarkId => ({
    type: 'API',
    payload: {
        url: `${BASE_URL}/api/delete`,
        id: bookmarkId,
    },
    actionType: DELETE_BOOKMARK_ACTION_TYPE,
});

export const upgradeBookmark = bookmarkId => ({
    type: 'API',
    payload: {
        url: `${BASE_URL}/api/upgrade`,
        id: bookmarkId,
    },
    actionType: UPGRADE_BOOKMARK_ACTION_TYPE,
});
