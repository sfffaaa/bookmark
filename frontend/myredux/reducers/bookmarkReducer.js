import { CREATE_BOOKMARK_ACTION_TYPE, LIST_BOOKMARK_ACTION_TYPE }
    from '../constants/action-types';

/* eslint-disable-next-line import/prefer-default-export */
export const bookmarkReducer = (bookmarks = {data: [], needUpdate: false}, action) => {
    switch (action.type) {
    case CREATE_BOOKMARK_ACTION_TYPE.PENDING:
        console.log(action);
        return bookmarks;
    case CREATE_BOOKMARK_ACTION_TYPE.ERROR:
        console.log(action);
        return bookmarks;
    case CREATE_BOOKMARK_ACTION_TYPE.SUCCESS:
        console.log(action);
        // [TODO] Add need update exist.
        return Object.assign({}, bookmarks, {
            data: bookmarks.data.concat({...action.response.data.data }),
            needUpdate: true,
        });
    case LIST_BOOKMARK_ACTION_TYPE.PENDING:
        console.log(action);
        return bookmarks;
    case LIST_BOOKMARK_ACTION_TYPE.ERROR:
        console.log(action);
        return bookmarks;
    case LIST_BOOKMARK_ACTION_TYPE.SUCCESS:
        console.log(action);
        return Object.assign({}, bookmarks, {
            data: action.response.data.data,
            needUpdate: false,
        });
    default:
        return bookmarks;
    }
};
