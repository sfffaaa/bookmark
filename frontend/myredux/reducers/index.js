import { combineReducers } from 'redux';
import { bookmarkReducer } from './bookmarkReducer';

/* eslint-disable-next-line import/prefer-default-export */
export const myCombineReducers = combineReducers({
    bookmarks: bookmarkReducer,
});
