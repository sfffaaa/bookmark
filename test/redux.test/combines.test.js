/* global describe it expect */
import deepFreeze from 'deep-freeze';

import * as actions from '../../frontend/myredux/actions/bookmarks';
import { bookmarkReducer } from '../../frontend/myredux/reducers/bookmarkReducer';
import { MYURL } from './constant';

describe('Redux action + reductor test', () => {
    it('should create an action to add a bookmark in empty list', () => {
        expect(bookmarkReducer([], actions.createBookmark(MYURL))).toMatchSnapshot();
    });

    it('should create an action to add a bookmark in empty list', () => {
        const initState = deepFreeze([{ url: 'http://my.test' }]);
        expect(bookmarkReducer(initState, actions.createBookmark(MYURL))).toMatchSnapshot();
    });
});
