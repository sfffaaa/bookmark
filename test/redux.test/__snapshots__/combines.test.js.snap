// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux action + reductor test should create an action to add a bookmark in empty list 1`] = `
Array [
  Object {
    "url": "http://NOUSEURL",
  },
]
`;

exports[`Redux action + reductor test should create an action to add a bookmark in empty list 2`] = `
Array [
  Object {
    "url": "http://my.test",
  },
  Object {
    "url": "http://NOUSEURL",
  },
]
`;
